feat: Adds file owner and group to the events' meta fields

Hiruma31 opened the pull request:
<!-- Type of change
Please label this PR with one of the following labels, depending on the scope of your change:
- Bug
- Enhancement
- Breaking change
- Deprecation
- Cleanup
- Docs
-->

## Proposed commit message

<!-- Mandatory
Explain here the changes you made on the PR.

Please explain:

- WHAT: patterns used, algorithms implemented, design architecture, message processing, etc.
- WHY:  the rationale/motivation for the changes

This text will be pasted into the squash dialog when the change is committed and will be
a long term historical record of the change to help future contributors understand the
change, please help them by making it clear and comprehensive, they may be you.

If the commit title is adequate to describe both of these things, The text here may be omitted
or replaced with "See title". The title of the PR will be used as the commit message title when
the merge is made and the "See title" marker will be removed if present.

The text here and the PR title will be subject to the PR review process.
-->

## Checklist

<!-- Mandatory
Add a checklist of things that are required to be reviewed in order to have the PR approved

List here all the items you have verified BEFORE sending this PR. Please DO NOT remove any item, striking through those that do not apply. (Just in case, strikethrough uses two tildes. ~~Scratch this.~~)
-->

- [x] My code follows the style guidelines of this project
- [x] I have commented my code, particularly in hard-to-understand areas
- [x] I have made corresponding changes to the documentation
- [x] I have made corresponding change to the default configuration files
- [x] I have added tests that prove my fix is effective or that my feature works. Where relevant, I have used the [`stresstest.sh`](https://github.com/elastic/beats/blob/main/script/stresstest.sh) script to run them under stress conditions and race detector to verify their stability.
- [x] I have added an entry in `./changelog/fragments` using the [changelog tool](https://github.com/elastic/elastic-agent-changelog-tool/blob/main/docs/usage.md).

## Disruptive User Impact

<!--
Will the changes introduced by this PR cause disruption to users in any way? If so, please describe what changes users
could make on their end to nullify or minimize this disruption. Consider impacts in related systems, not just directly
when using Beats.
-->
No disruption, only a slightly larger event payload. Error catching is handled the exact same way as other fields.

## Author's Checklist

<!-- Recommended
Add a checklist of things that are required to be reviewed in order to have the PR approved
-->
- [ ]

## How to test this PR locally

<!-- Recommended
Explain here how this PR will be tested by the reviewer: commands, dependencies, steps, etc.
-->

- Build this version
- Runs the generated Filebeat on a Linux host and with a filestream input
- Observe the event generated in the output, it will include user and group fields
```
{
[...]
  "log": {
    "file": {
      "group": "mgroup",
      "path": "/tmp/test.log",
      "device_id": "426879",
      "inode": "1292",
      "owner": "myuser"
    }
  }
[...]
}

```

## Related issues

<!-- Recommended
Link related issues below. Insert the issue link or reference after the word "Closes" if merging this should automatically close it.

- Closes #123
- Relates #123
- Requires #123
- Superseds #123
-->
- Closes #43226

## Use cases

<!-- Recommended
Explain here the different behaviors that this PR introduces or modifies in this project, user roles, environment configuration, etc.

If you are familiar with Gherkin test scenarios, we recommend its usage: https://cucumber.io/docs/gherkin/reference/
-->

## Screenshots

<!-- Optional
Add here screenshots about how the project will be changed after the PR is applied. They could be related to web pages, terminal, etc, or any other image you consider important to be shared with the team.
-->

## Logs

<!-- Recommended
Paste here output logs discovered while creating this PR, such as stack traces or integration logs, or any other output you consider important to be shared with the team.
-->

Hiruma31 commented:
I have signed the CLA, not sure why this test fails.

cmacknz commented:
I think this should be configurable and disabled by default, as suggested in the original issue. 

All of this metadata ideally would be opt in, with the existing fields enable by default and these new ones disable by default to avoid a breaking change to the metadata. At least in the scope of this PR, the new fields should be opt in. Following the conventions used by [filelog](https://github.com/open-telemetry/opentelemetry-collector-contrib/tree/main/receiver/filelogreceiver) with `include_file_owner_name` and `include_file_owner_group_name` seem reasonable to me as names.

This metadata increases the storage overhead of every log file we ingest so unless you are specifically trying to query the information it provides, it just drives up the storage cost of logs.

belimawr reviewed (commented):
Overall the PR looks good, thanks for the contribuition @Hiruma31 !

However I agree with @cmacknz, this should be disabled by default.

belimawr commented:
> @belimawr, do u agree?

Yes, that sounds good. Thanks @AndersonQ!

belimawr commented on filebeat/filebeat.reference.yml:
```diff
@@ -861,6 +861,10 @@ filebeat.inputs:
 
+  # Include file owner name and group name in the event metadata.
+  #include_file_owner_name: true
+  #include_file_owner_group_name: true
```
A couple of things here:
1. in the reference documentation we usually use the default values, so those should be `false`
2. This file is auto-generated based on [`filebeat/_meta/config/filebeat.inputs.reference.yml.tmpl`](https://github.com/elastic/beats/blob/main/filebeat/_meta/config/filebeat.inputs.reference.yml.tmpl).

To generate the auto-generated files:
```
cd filebeat
mage check
cd ../x-pack/filebeat
mage check
```

`mage check` does more than just generating the files, it:
- Formats the code
- Generate all files
- Checks whether there are uncommitted changes.

You need to run this on both `filebeat` and `x-pack/filebeat` as the latter uses the config templates from former.

belimawr reviewed (commented):
One last thing that would be nice, but might be a bit hard to get started: adding an integration test.

We have a number of integration tests on [`filebeat/tests/integration`](https://github.com/elastic/beats/tree/main/filebeat/tests/integration). The use our [integration tests framework](https://github.com/elastic/beats/tree/main/libbeat/tests/integration).

You can use [this one](https://github.com/elastic/beats/blob/ed6848fefe22ef50a2b0b84ab1bb744949e19598/filebeat/tests/integration/filestream_test.go#L80-L126) as a starting point.

You'd setup Filebeat using the file output, ingest the data and check whether the events contain the correct new fields. The framework has a function to get events from file output: [`GetEventsFromFileOutput`](https://github.com/elastic/beats/blob/ed6848fefe22ef50a2b0b84ab1bb744949e19598/libbeat/tests/integration/framework.go#L1304).

Let me know if you need any help with the integration tests, I can help you out if needed.

Hiruma31 commented:
> This is a new feature, it should not be backported, only bug fixes are always backported.
> 
> Are you in need of this feature in 8.19?

That would be ideal, or at least in Filebeat 8, I'm not too concerned with the minor. Maybe the label should be `backport-active-8` ? I am not sure if you run 2 major concurrently.

I'll address the rest of the comments tonight or tomorrow, thanks a lot for the detailed review @belimawr ! :)

belimawr commented:
> That would be ideal, or at least in Filebeat 8, I'm not too concerned with the minor. Maybe the label should be `backport-active-8` ? I am not sure if you run 2 major concurrently.

We do not run 2 majors concurrently, we keep the last minor from the last major (8.19 in this case) receiving some bug and security fixes, but we do not add new features to it :/

This code will be released in the next 9.x minor after it is merged.

If you really, really need it in a 8.19 build, you can build it from your fork.

kruskall commented:
> I have signed the CLA, not sure why this test fails.

@Hiruma31 The CLA was signed but the 3 of the 4 commits in this PR are not associated with your account (ie. the username is set but the email is empty).

AndersonQ commented on libbeat/reader/readfile/metafields_windows_test.go:
```diff
@@ -70,3 +70,27 @@ func checkFields(t *testing.T, expected, actual mapstr.M) {
+	require.NoError(t, err)
+
+	require.Equal(t, expected, actual)
+}
```
This is the same as `checkFields` above. If they ca be the same, just call `checkFields` inside `checkFieldsWithOwnerGroup`


```suggestion
func checkFieldsWithOwnerGroup(t *testing.T, expected, actual mapstr.M) {
	checkFields(t, expected, actual)
}
```

belimawr replied:
+1

I'd also add comment explaining that on Windows we don't support adding owner and group, hence it only calls `checkFields`

belimawr commented on filebeat/tests/integration/filestream_test.go:
```diff
@@ -678,6 +678,50 @@ func TestFilestreamTakeOverFromLogInput(t *testing.T) {
+		Message string `json:"message"`
+	}
+
+	evts := integration.GetEventsFromFileOutput[evt](filebeat, 5, false)
```
You will also need to set the `filename` as it defaults to `filebeat` and the function expects `output`. Here is an [example](filebeat/tests/integration/testdata/filebeat_journald.yml) of what to configure to use the file output with debug logging:
```yaml
path.home: %s

queue.mem:
  flush.timeout: 0

output:
  file:
    path: ${path.home}
    filename: "output"
    rotate_on_startup: false

logging:
  level: debug
  selectors:
    - "*"
```

And the code using it to configure Filebeat:
https://github.com/elastic/beats/blob/be9871f5dab9888ab7b56d96567e9016a17d2ec8/filebeat/tests/integration/journald_test.go#L50-L55

belimawr commented on libbeat/reader/readfile/metafields_test.go:
```diff
@@ -80,6 +80,91 @@ func TestMetaFields(t *testing.T) {
+					in.checkFieldsMethod(t, expectedFields, msg.Fields)
+				} else {
+					require.Equal(t, expectedFields, msg.Fields)
+				}
```
I'm a bit confused here on why you're also testing the case where the fields are empty. I get you were inspired by the test before it, but because this is only testing the features added by this PR we can simplify it:
1. Remove the empty message from `message`
2. Remove the `len(msg.Content) != 0` statement.

```suggestion
				expectedFields := mapstr.M{
					"log": mapstr.M{
						"file": mapstr.M{
							"path":        path,
							"fingerprint": "hash",
						},
						"offset": offset,
					},
				}
				in.checkFieldsMethod(t, expectedFields, msg.Fields)
```

belimawr commented on docs/reference/filebeat/filebeat-input-filestream.md:
```diff
@@ -216,6 +216,16 @@ A list of glob-based paths that will be crawled and fetched. All patterns suppor
+
+Includes the log file group to `log.file` metadata.
+
+
```
You need to expand a bit explaining this feature does not work on Windows.

Hiruma31 replied:
I reused a simple "This option is not supported on Windows". I'm still considering bringing that support in a follow-up PR so I didn't want to put too much details as to why it's not a 1:1 situation.

belimawr commented on filebeat/tests/integration/filestream_test.go:
```diff
@@ -678,6 +678,50 @@ func TestFilestreamTakeOverFromLogInput(t *testing.T) {
 	}
 }
 
+func TestFilestreamHasOwnerAndGroup(t *testing.T) {
```
Another important thing, the tests on this file also run on Windows, but this feature is not supported on Windows.

My recommendation is to put it on a separate file and add the `!windows` build flag:
```go
//go:build integration && !windows
```

Hiruma31 commented on libbeat/reader/readfile/metafields_test.go:
```diff
@@ -80,6 +80,91 @@ func TestMetaFields(t *testing.T) {
 	}
 }
 
+func TestMetaFieldsNew(t *testing.T) {
```
My bad that's actually an unsaved typo. I'll use your suggestion.

belimawr commented on filebeat/tests/integration/filestream_other_test.go:
```diff
@@ -0,0 +1,100 @@
+	tempDir := filebeat.TempDir()
+	logFilePath := filepath.Join(tempDir, "input.log")
+
+	integration.WriteLogFile(t, logFilePath, 5, false)
```
Only 5 lines will generate a file with only 250 bytes, which is not large enough to be ingested by Filestream with the default file identity from 9.0.

You need to update it to at least 25 lines:
```suggestion
	integration.WriteLogFile(t, logFilePath, 25, false)
```

belimawr commented on docs/reference/filebeat/filebeat-input-filestream.md:
```diff
@@ -216,6 +216,18 @@ A list of glob-based paths that will be crawled and fetched. All patterns suppor
+
+Includes the log file group to `log.file` metadata. 
+This option is not supported on Windows.
+
```
We use cumulative docs now, this means there is a single page/version of the documentation to all 9.x released.

So each new feature needs to be tagged with the correct version, for that you need to add the `applies_to` badge:

\`\`\`yaml {applies_to}
stack: ga 9.3
\`\`\`


Here are the docs explaining them: https://elastic.github.io/docs-builder/syntax/applies/#section-level

Sorry for not catching that earlier 🙈

Hiruma31 replied:
No problem I added it to both sections as requested

Hiruma31 commented on filebeat/tests/integration/filestream_other_test.go:
```diff
@@ -0,0 +1,100 @@
+    path: ${path.home}
+    filename: "output"
+    rotate_on_startup: false
+`, logFilePath)
```
Tabulations instead of spaces... 🤦‍♂️

AndersonQ commented on docs/reference/filebeat/filebeat-input-filestream.md:
````diff
@@ -216,6 +216,24 @@ A list of glob-based paths that will be crawled and fetched. All patterns suppor
+```
+
+Includes the log file group to `log.file` metadata. 
+This option is not supported on Windows.
````
I was checking the rendered docs, and I believe it's swapped:


```suggestion
```yaml {applies_to}
stack: ga 9.3
```
#### `include_file_owner_name` [filestream-input-include_file_owner_name]

Includes the log file owner to `log.file` metadata. 
This option is not supported on Windows.

```yaml {applies_to}
stack: ga 9.3
```
#### `include_file_owner_group_name` [filestream-input-include_file_owner_group_name]

Includes the log file group to `log.file` metadata. 
This option is not supported on Windows.
```

Hiruma31 commented on docs/reference/filebeat/filebeat-input-filestream.md:
```diff
@@ -216,6 +216,24 @@ A list of glob-based paths that will be crawled and fetched. All patterns suppor
+Includes the log file group to `log.file` metadata. 
+This option is not supported on Windows.
+
+
```
No worries, understood and moved.

AndersonQ commented:
/test

AndersonQ reviewed (changes requested):
https://github.com/elastic/beats/pull/47331#issuecomment-3496097477 needs to be addressed to remove the unintended change in behaviour on `inodeDeviceIdentifier`. That's why the tests are failing

Hiruma31 commented:
@AndersonQ Sure thing, it's done. Although I have a doubt I would like to clarify since the patch doesn't seem to match exactly what I have. `StateOS.String()` used to be
```
func (fs StateOS) String() string {
	var buf [64]byte
	current := strconv.AppendUint(buf[:0], fs.Inode, 10)
	current = append(current, '-')
	current = strconv.AppendUint(current, fs.Device, 10)
	return string(current)
}
```
(your new `StringIgnodeDevice()`) whereas I added the UID and GID. 
In the patch I see you have just 
```
func (fs StateOS) String() string {
	return string(current)
}
```

Perhaps that's what i should do instead: simply revert what I had modified in the `String()` method ?

AndersonQ commented:
/test

AndersonQ commented:
hey @Hiruma31, 
I hope you don't mind, but I took the liberty of pushing the fix to your branch.
I should pass the tests now :)

Hiruma31 commented on libbeat/common/file/file_other.go:
```diff
@@ -56,6 +60,22 @@ func (fs StateOS) IsSame(state StateOS) bool {
+// Filebeat inodeDeviceIdentifier uses this string when generating its ID. With
+// the addition of UID and GID in the String() method, we need to keep the old
+// behaviour for backward compatibility.
+func (fs StateOS) StringInodeDevice() string {
```
Makes sense, sorry for the delay, I renamed the method to Identifier() everywhere it was called.

cmacknz commented:
> @cmacknz do you think we need a benchmark for that feature?
> It's disable by default, so not issue for current users. I just did a quick one here there > is a difference. But The question is if it'll have a real impact on the whole.

You are correct that it is disabled by default so it is not a concern for the main path. This is at least two additional system calls for each line we read from a file so it should show up in benchmarks to some degree.

This could probably be solved with caching or something simpler like only reading the owner and group when you open the file instead of on every line read from the file so it's not like it's unsolvable.

AndersonQ commented:
issue created: 
 - https://github.com/elastic/beats/issues/47878

belimawr reviewed (approved):
LGTM, just a small comment about adding the `id` for Filestream in the test config. But it's non-blocking.

AndersonQ commented:
/test

colleenmcginnis reviewed (approved):
Docs updates look good. 🙂

AndersonQ commented:
/test
